### PR TITLE
Compress aggregator placement

### PR DESCRIPTION
**Pull request by abliqo**
**Problem:**

Large deployments of M3 aggregator of 1K+ aggregator nodes and O(10K) clients experience scale issues when placement changes:
- the size of placement is a function of number of aggregator nodes and shards
- the total amount of placement data that etcd has to propagate to all clients becomes very large (`payload_size * N_clients`)
- upon placement change, etcd gets overloaded and struggles to propagate so much data to all clients
- clients that don't get updated placement use stale version which increases the risk of data loss when aggregator nodes are replaced and requires manual intervention

**Solution:**

If placement was compressed, its size (protobuf-encoded) could be reduced by ~4x times.

This change introduces opt-in compression of the aggregator placement in backward compatible way:
- introduce new option `SetCompress` that determines whether the placement should be compressed when it is written to etcd
- that option is `false` by default to not break existing clients
- the code reading the placement from etcd handles both uncompressed and compressed versions

All clients, including aggregator nodes, must be upgraded before compression option can be enabled in placement management code.

**Comparison of compression libraries**

Using Go's standard `compress/flate` with best compression option:
```
BenchmarkCompressPlacementProto-12      	      10	 107019136 ns/op	         3.76 compress_rate/op
BenchmarkDecompressPlacementProto-12    	     171	   6995907 ns/op
```

Using `zstd` with best compression option:
```
BenchmarkCompressPlacementProto-12      	      37	  28166644 ns/op	         4.00 compress_rate/op
BenchmarkDecompressPlacementProto-12    	     172	   7214039 ns/op
```

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/cluster/placement/compress.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package placement
+
+import (
+	"bytes"
+	"compress/flate"
```
Comment (vdarulis) [inline src/cluster/placement/compress.go:25]: can we use https://github.com/klauspost/compress instead?

Comment (abliqo) [reply src/cluster/placement/compress.go:25]: Could you please elaborate why? :)

Comment (vdarulis) [reply src/cluster/placement/compress.go:25]: more flexible api's (more interfaces generated for you), faster, drop-in replacement

File: go.mod [hunk 2]
```diff
@@ -73,6 +73,7 @@ require (
 	github.com/prometheus/client_golang v1.5.1
 	github.com/prometheus/common v0.9.1
 	github.com/prometheus/prometheus v1.8.2-0.20200420081721-18254838fbe2
+	github.com/quasilyte/go-ruleguard/dsl/fluent v0.0.0-20201222093424-5d7e62a465d3 // indirect
```
Comment (vdarulis) [inline go.mod:76]: you need to run `go mod tidy`

File: src/cluster/placement/compress.go (hunk 1 above)
Comment (abliqo) [reply src/cluster/placement/compress.go:25]: I updated benchmark test to use large simulated randomized placement:
```
pkg: github.com/m3db/m3/src/cluster/placement
BenchmarkCompressPlacementProto-12      	      10	 107019136 ns/op	         3.76 compress_rate/op
BenchmarkDecompressPlacementProto-12    	     171	   6995907 ns/op
```

Given how insignificant is compression/decompression overhead time, I would rather stick to standard library for the sake of stability.

Comment (abliqo) [reply src/cluster/placement/compress.go]: switched to that library and using `zstd` - it is indeed slightly better, thanks for suggestion

File: src/cluster/placement/compress.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+	uncompressed, _ := p.Marshal()
+	opts := zstd.WithEncoderLevel(zstd.SpeedBestCompression)
+	var compressed bytes.Buffer
+	w, err := zstd.NewWriter(&compressed, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err := w.Write(uncompressed); err != nil {
```
Comment (vdarulis) [inline src/cluster/placement/compress.go:46]: use `w.EncodeAll` for even better compression ratio

File: src/cluster/placement/compress.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+	return compressed.Bytes(), nil
+}
+
+func decompressPlacementProto(compressed []byte) (*placementpb.Placement, error) {
+	if compressed == nil {
+		return nil, errNilValue
+	}
+
+	r, err := zstd.NewReader(bytes.NewReader(compressed))
```
Comment (vdarulis) [inline src/cluster/placement/compress.go:63]: use `w.DecodeAll` to decode from/to a raw buffer

File: src/cluster/placement/compress.go (hunk 3 above)
Comment (abliqo) [reply src/cluster/placement/compress.go]: Updated, although it makes virtually no difference: 4.01 vs 4.00 compression rate.

Comment (abliqo) [reply src/cluster/placement/compress.go]:
```
BenchmarkCompressPlacementProto-12      	      30	  36173342 ns/op	         4.01 compress_rate/op
BenchmarkDecompressPlacementProto-12    	     213	   5722368 ns/op
```

File: src/cluster/generated/proto/placementpb/placement.proto [hunk 5]
```diff
@@ -76,6 +76,14 @@ enum ShardState {
   LEAVING = 2;
 }
 
+enum CompressMode {
+  NONE = 0;
+  FLATE = 1;
```
Comment (vdarulis) [inline src/cluster/generated/proto/placementpb/placement.proto:81]: s/FLATE/ZSTD/?

**Review by vdarulis — APPROVED** [review]
lgtm w/ comment

File: src/cluster/generated/proto/placementpb/placement.proto (hunk 5 above)
Comment (abliqo) [reply src/cluster/generated/proto/placementpb/placement.proto:81]: Oh, of course, good catch!

**Review by vdarulis — APPROVED** [review]

**Pull request merged**